squire_sdk/src/api/request.rs

```diff
@@ -0,0 +1,146 @@
+#[allow(unused_imports)]
+use std::convert::Infallible;
+
+#[cfg(feature = "server")]
+use axum::{body::HttpBody, handler::Handler, routing::MethodRouter};
+use serde::{de::DeserializeOwned, Serialize};
+
+use super::url::Url;
+#[cfg(feature = "server")]
+use crate::server::state::ServerState;
+
+/// An enum that is used by the `RestRequest` to "know" what `MethodRouter` contructor to call
+pub enum Method {
+    Get,
+    Post,
+    Patch,
+    Delete,
+}
+
+// Only integers, chars, and bools can be used as const generics, so we define consts for each
+// method variant.
+pub const GET: u8 = Method::Get as u8;
+pub const POST: u8 = Method::Post as u8;
+pub const PATCH: u8 = Method::Patch as u8;
+pub const DELETE: u8 = Method::Delete as u8;
+
+/// A function used to turn const-generics into Methods at compile time.
+///
+/// NOTE: This panics only at compile time, but it will do so when you run `cargo build` not `cargo
+/// check`.
+const fn to_method<const M: u8>() -> Method {
+    match M {
+        GET => Method::Get,
+        POST => Method::Post,
+        PATCH => Method::Patch,
+        DELETE => Method::Delete,
+        _ => panic!("Invalid method value"),
+    }
+}
+
+/// A trait that abstracts the connections between the data used in API definitions on the server's
+/// side. This trait is not intended to be implmented directly. Instead, a type should implement
+/// one of the method-specific traits since each of those has a blanket implementation for this
+/// trait.
+// TODO: Remove the generic `N` here. This trait does not need to have a full URL, only the string.
+pub trait RestRequest<const N: usize, const M: u8>: Serialize + DeserializeOwned {
+    const ROUTE: Url<N>;
+    const METHOD: Method = to_method::<M>();
+    type Response: DeserializeOwned;
+
+    // TODO: Currently, this method can not ensure that the return value of the handler is the same
+    // as the reponse type. We should support to have an even tighter link between client and
+    // server. Moveover, we should also support the encoding method of the handler.
+    #[cfg(feature = "server")]
```

Comment:
Yes, that is accurate, but that's ok. We only need this trait while constructing the server's router. After that point, we never use this data again. This will also be true for the features that have been discussed for the future.